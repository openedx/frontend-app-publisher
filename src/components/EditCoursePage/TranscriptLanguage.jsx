import React from 'react';
import PropTypes from 'prop-types';
import { Field, FieldArray } from 'redux-form';
import { Alert } from '@edx/paragon';

import RenderSelectField from '../RenderSelectField';
import RemoveButton from '../RemoveButton';
import FieldLabel from '../FieldLabel';

class TranscriptLanguage extends React.Component {
  constructor(props) {
    super(props);
    this.handleRemove = this.handleRemove.bind(this);
  }

  handleRemove(index) {
    this.props.fields.remove(index);
  }

  render() {
    const {
      fields,
      languageOptions,
      disabled,
      meta: {
        submitFailed,
        error,
      },
    } = this.props;

    return (
      // Set tabindex to -1 to allow programmatic shifting of focus for validation
      <div className="transcript-languages mb-3" name={fields.name} tabIndex="-1">
        {submitFailed && error
          && (
          <Alert variant="danger">{error}</Alert>
          )}
        <ul className="list-group p-0 m-0 container-fluid">
          {fields.map((language, index) => (
            <li className="transcript-language list-group-item row d-flex align-items-center px-0 mx-0" key={language}>
              <div className="col-11">
                <Field
                  name={`${language}`}
                  component={RenderSelectField}
                  options={languageOptions}
                  type="text"
                  label={<FieldLabel text="Transcript language" required />}
                  extraInput={{ onInvalid: this.openCollapsible }}
                  disabled={disabled}
                  required
                />
              </div>
              <RemoveButton
                className="col-1"
                label="Remove language"
                onRemove={this.handleRemove}
                targetFieldNumber={index}
                disabled={disabled}
              />
            </li>
          ))}
        </ul>
        <button
          type="button"
          className="btn btn-outline-primary js-add-button mt-2"
          onClick={() => fields.push({})}
          disabled={disabled}
        >
          Add language
        </button>
      </div>
    );
  }
}

TranscriptLanguage.propTypes = {
  fields: PropTypes.instanceOf(FieldArray).isRequired,
  languageOptions: PropTypes.arrayOf(PropTypes.shape({
    label: PropTypes.string,
    value: PropTypes.string,
  })).isRequired,
  disabled: PropTypes.bool,
  extraInput: PropTypes.shape({}),
  meta: PropTypes.shape({
    submitFailed: PropTypes.bool,
    error: PropTypes.string,
  }).isRequired,
};

TranscriptLanguage.defaultProps = {
  disabled: false,
  extraInput: {},
};

export default TranscriptLanguage;
