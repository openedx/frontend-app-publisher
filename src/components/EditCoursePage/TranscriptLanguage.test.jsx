import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import TranscriptLanguage from './TranscriptLanguage';

/*
*  Disable console errors for this test file so that we don't receive warnings
*  about fields being an array rather than an object. This prop change is
*  intentional as redux-form field arrays treat fields strangely:
*  'The fields object is a "pseudo-array", in that it has many of the same
*  properties and methods as a javascript Array, providing both reading and
*  writing functionality.'
*/
jest.spyOn(global.console, 'error').mockImplementation(() => jest.fn());

const languageOptions = [{
  label: 'Arabic - United Arab Emirates',
  value: 'ar-ae',
}];

const meta = {
  submitFailed: false,
  error: '',
};

const metaFailed = {
  submitFailed: true,
  error: 'This field is required',
};

describe('Transcript Language', () => {
  it('renders correctly with no fields', () => {
    const component = shallow(<TranscriptLanguage
      fields={[]}
      languageOptions={languageOptions}
      meta={meta}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('renders correctly when given fields', () => {
    const component = shallow(<TranscriptLanguage
      fields={[{}]}
      languageOptions={languageOptions}
      meta={meta}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('renders correctly with an error after failed submission', () => {
    const component = shallow(<TranscriptLanguage
      fields={[{}]}
      languageOptions={languageOptions}
      meta={metaFailed}
    />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('adds fields when the add button is pushed', () => {
    const fields = [{}];
    const component = shallow(<TranscriptLanguage
      fields={fields}
      languageOptions={languageOptions}
      meta={meta}
    />);
    expect(fields.length).toEqual(1);

    component.find('.js-add-button').simulate('click');
    expect(fields.length).toEqual(2);
  });
});
