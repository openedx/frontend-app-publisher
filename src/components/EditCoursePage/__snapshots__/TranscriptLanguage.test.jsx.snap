// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transcript Language renders correctly when given fields 1`] = `
<div
  className="transcript-languages mb-3"
  tabIndex="-1"
>
  <ul
    className="list-group p-0 m-0 container-fluid"
  >
    <li
      className="transcript-language list-group-item row d-flex align-items-center px-0 mx-0"
      key="[object Object]"
    >
      <div
        className="col-11"
      >
        <Field
          component={[Function]}
          disabled={false}
          extraInput={
            Object {
              "onInvalid": undefined,
            }
          }
          label={
            <FieldLabel
              className=""
              extraText=""
              helpText=""
              id={null}
              optional={false}
              required={true}
              text="Transcript language"
            />
          }
          name="[object Object]"
          options={
            Array [
              Object {
                "label": "Arabic - United Arab Emirates",
                "value": "ar-ae",
              },
            ]
          }
          required={true}
          type="text"
        />
      </div>
      <RemoveButton
        className="col-1"
        disabled={false}
        label="Remove language"
        onRemove={[Function]}
        targetFieldNumber={0}
      />
    </li>
  </ul>
  <button
    className="btn btn-outline-primary js-add-button mt-2"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Add language
  </button>
</div>
`;

exports[`Transcript Language renders correctly with an error after failed submission 1`] = `
<div
  className="transcript-languages mb-3"
  tabIndex="-1"
>
  <StatusAlert
    alertType="danger"
    className=""
    dismissible={false}
    iconClassNames={Array []}
    message="This field is required"
    onClose={[Function]}
    title={null}
  />
  <ul
    className="list-group p-0 m-0 container-fluid"
  >
    <li
      className="transcript-language list-group-item row d-flex align-items-center px-0 mx-0"
      key="[object Object]"
    >
      <div
        className="col-11"
      >
        <Field
          component={[Function]}
          disabled={false}
          extraInput={
            Object {
              "onInvalid": undefined,
            }
          }
          label={
            <FieldLabel
              className=""
              extraText=""
              helpText=""
              id={null}
              optional={false}
              required={true}
              text="Transcript language"
            />
          }
          name="[object Object]"
          options={
            Array [
              Object {
                "label": "Arabic - United Arab Emirates",
                "value": "ar-ae",
              },
            ]
          }
          required={true}
          type="text"
        />
      </div>
      <RemoveButton
        className="col-1"
        disabled={false}
        label="Remove language"
        onRemove={[Function]}
        targetFieldNumber={0}
      />
    </li>
  </ul>
  <button
    className="btn btn-outline-primary js-add-button mt-2"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Add language
  </button>
</div>
`;

exports[`Transcript Language renders correctly with no fields 1`] = `
<div
  className="transcript-languages mb-3"
  tabIndex="-1"
>
  <ul
    className="list-group p-0 m-0 container-fluid"
  />
  <button
    className="btn btn-outline-primary js-add-button mt-2"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    Add language
  </button>
</div>
`;
